### DISCUSSION TITLE
Idea: Enable and Disable ipp-usb

### DISCUSSION THREAD

**Discussion by martin734** [OpenPrinting/ipp-usb#35] [Ideas]
A printer/scanner obviously has 2 modes of operation:

1. Printer
2. Scanner

The Linux Mint User Guide has the following warning:

> When ipp-usb is installed it takes over communication with all printing/scanning devices, i.e. printing/scanning drivers cannot work, they’re inhibited.

As a consequence ipp-usb must be disabled for printing.
Question: How can ipp-usb enabled and disabled on a running system?
I would like to implement 2 scripts `enablePrinter.sh` and `enableScanner.sh`. This might be implemented if ipp-usb supports the following enhancement:

Activate and deactivate ipp-usb with a signal

It might also be useful to query if ipp-usb is enabled or disabled.

Or is this functionality already built in and I missed it?
Thank you

**Comment by alexpevzner** [comment]
On Linux Mint `ipp-usb` is controlled by `systemd` and standard methods to disable/enable service should work:

```
systemctl disable ipp-usb
systemctl enable ipp-usb
```

Note, dynamic switch between IPP over USB and "legacy" modes of operation known to work unstable on some devices (for example, after attempt to use "legacy" USB printing interface, IPP over USB may stop working until device reboot). So I don't recommend this approach.

Printing should work via ipp-usb as well as scanning does. If there are some problems with printing, it is better to fix these problems that switch device access mode dynamically.

**Reply by martin734** [reply]
Systemctl does not work:
`$sudo systemctl enable ipp-usb
The unit files have no installation config (WantedBy=, RequiredBy=, Also=, 
Alias= settings in the [Install] section, and DefaultInstance= for template
units). This means they are not meant to be enabled using systemctl.`

 
Possible reasons for having this kind of units are:
• A unit may be statically enabled by being symlinked from another unit's
  .wants/ or .requires/ directory.
• A unit's purpose may be to act as a helper for some other unit which has
  a requirement dependency on it.
• A unit may be started when needed via activation (socket, path, timer,
  D-Bus, udev, scripted systemctl call, ...).
• In case of template units, the unit is meant to be enabled with some
  instance name specified.

Maybe the installation config is intentionally not in the installation package because ipp-usb should not be switched on and off.

**Comment by alexpevzner** [comment]
Probably, @tillkamppeter can help with fixing problems at printing side.

**Comment by tillkamppeter** [comment]
If you have a device which supports IPP-over-USB, in IPP-over-USB mode (with active ipp-usb) the device works driverless for all its functionality: Printing, scanning, fax out, admin interface. So you should not need to switch between IPP-over-USB and classic USB input/output stream mode.

Important is that you print in driverless mode, not using PPD files of HPLIP, Gutenprint, or whatever printer driver. Also do not use the Printer Applications (PostScript, Ghostscript, HPLIP, Gutenprint) offered in the Snap Store (they are printer drivers). IPP-over-USB is driverless-only.

CUPS should create print queues automatically, or cups-browsed does, or you select the "driverless" entries in your printer setup tool (under network printers, as ipp-usb emulates an IPP network printer).

**Reply by martin734** [reply]
Thank you, it works.
You write 

> the device works driverless for all its functionality


However some functionality  is lacking, for instance 

- Print Quality: Only "Normal" and "High", in TurboPrint there is also a draft-mode
- Print Color Mode: Only "Color" and "Monochrome", in TurboPrint there is also a "Greyscale"
- Layout and color-setting seems not possible

That's no big deal, I certainly can cope with that.
I have still another question: is there a possibility to store different configurations which could be selected in LibreOffice? 

**Again, thank all of you very much**. To find a solution to problems of TurboPrint has kept me busy for several weeks now. Therefore I am glad that printing works without this software.

**Comment by alexpevzner** [comment]
@tillkamppeter,

> Important is that you print in driverless mode, not using PPD files of HPLIP, Gutenprint, or whatever printer driver

honestly speaking, it doesn't always work. For example, my Kyocera ECOSYS M2040dn doesn't print with driverless driver automatically installed. To get printing work, I had to manually extract PPD file from the proprietary driver and to manually tell CUPS to use it. Interesting, proprietary driver also doesn't work on recent Fedora, where Python 2 is depricated (proprietary driver depends on it).
